fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OutputPropertyDeriver.java

```diff
@@ -0,0 +1,576 @@
+
+        // update group expression require property
+        PhysicalPropertySet newRightChildInputProperty = createPropertySetByDistribution(rightDistributionSpec);
+        requiredChildrenProperties.set(1, newRightChildInputProperty);
+    }
+
+    // enforce child SHUFFLE type distribution
+    private void enforceChildShuffleDistribution(List<Integer> shuffleColumns, GroupExpression child,
+                                                 PhysicalPropertySet childOutputProperty, int childIndex) {
+        DistributionSpec enforceDistributionSpec =
+                DistributionSpec.createHashDistributionSpec(new HashDistributionDesc(shuffleColumns,
+                        HashDistributionDesc.SourceType.SHUFFLE_ENFORCE));
+
+        enforceChildDistribution(enforceDistributionSpec, child, childOutputProperty);
+
+        PhysicalPropertySet newChildInputProperty = createPropertySetByDistribution(enforceDistributionSpec);
+        requiredChildrenProperties.set(childIndex, newChildInputProperty);
+    }
+
+    private void enforceChildDistribution(DistributionSpec distributionSpec, GroupExpression child,
+                                          PhysicalPropertySet childOutputProperty) {
+        double childCosts = child.getCost(childOutputProperty);
+        if (child.getOp() instanceof PhysicalDistributionOperator) {
+            // update child distribution directly
+            PhysicalDistributionOperator rightChildOperator = (PhysicalDistributionOperator) child.getOp();
+            rightChildOperator.setDistributionSpec(distributionSpec);
+            // update child group expression low cost table
+            PhysicalPropertySet newChildOutputProperty = createPropertySetByDistribution(distributionSpec);
+
+            child.setPropertyWithCost(newChildOutputProperty,
+                    child.getInputProperties(childOutputProperty), childCosts);
+            child.getGroup()
+                    .setBestExpression(child, childCosts, newChildOutputProperty);
+            if (ConnectContext.get().getSessionVariable().isSetUseNthExecPlan()) {
+                // record the output/input properties when child group could satisfy this group expression required property
+                child.addValidOutputInputProperties(newChildOutputProperty,
+                        child.getInputProperties(childOutputProperty));
+                child.getGroup().addSatisfyRequiredPropertyGroupExpression(newChildOutputProperty, child);
+            }
+        } else {
+            // add physical distribution operator
+            addChildEnforcer(childOutputProperty, new DistributionProperty(distributionSpec),
+                    childCosts, child.getGroup());
+        }
+    }
+
+    // compute the physical property info, just compute the nullable columns now
+    public PhysicalPropertySet computeHashJoinPhysicalPropertyInfo(PhysicalHashJoinOperator node,
```

Comment:
nullable info is logic property, not physical property.

Reply:
renamed